### PR TITLE
AA-HCA Preemption Policy

### PR DESCRIPTION
**Pull request by alessandroliafook**
Implementing AA-HCA Preemption Policy, and the choice between options of allocation policy and preemption policy

### CODE REVIEW

File: nova/scheduler/weights/worst_fit.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+Worst Fit Weigher.  Weigh hosts by their capacity.
+
+"""
+
+import nova.conf
+from nova.scheduler import weights
+from nova.scheduler.preemption_policy import PreemptionPolicy
+
+CONF = nova.conf.CONF
+
```
Comment (fabiomorais) [inline nova/scheduler/weights/worst_fit.py]: Please, remove empty lines and unnecessary prints.

File: nova/scheduler/weights/best_fit.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+Worst Fit Weigher.  Weigh hosts by their capacity.
+
+"""
+
+import nova.conf
+from nova.scheduler import weights
+from nova.scheduler.preemption_policy import PreemptionPolicy
+
+CONF = nova.conf.CONF
+
```
Comment (fabiomorais) [inline nova/scheduler/weights/best_fit.py]: Please, remove empty lines and unnecessary prints.

File: nova/scheduler/weights/best_fit.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+class BestFitWeigher(weights.BaseHostWeigher):
+
+    def _weigh_object(self, host_state, weight_properties):
+
+        print "Weighing..."
+        preemption_policy = PreemptionPolicy()
+        slo_new_request = float(weight_properties.get_scheduler_hint('slo'))
+        print "Slo of the request is %d" % slo_new_request
+        free_vcpu = preemption_policy.get_host_capacity_by_slo(host_state, weight_properties)
+        value = float(host_state.vcpus_total - free_vcpu)
```
Comment (fabiomorais) [inline nova/scheduler/weights/best_fit.py]: Should be this line inside the preemption policy?

File: nova/scheduler/preemptive_filter_scheduler.py [hunk 4]
```diff
@@ -163,8 +164,22 @@ def _schedule(self, context, spec_obj):
 
             print "Weighed hosts:"
             print weighed_hosts
+            
+            if len(weighed_hosts) > 1:
```
Comment (fabiomorais) [inline nova/scheduler/preemptive_filter_scheduler.py]: Please, use list comprehensions to be more concise.

File: nova/scheduler/preemption_policy/fcfs_preemption_policy.py [hunk 5]
```diff
@@ -26,3 +26,19 @@ def compare_vms(spec_obj1, spec_obj2):
...
+def compare_vms_to_preempt(spec_obj1, spec_obj2):
+    return _compare_vms(spec_obj1, spec_obj2)
+
+
+def preempt_vm(new_request_spec, spec_running):
+
+    slo_new_request = float(new_request_spec.get_scheduler_hint('slo'))
+    slo_running_request = float(spec_running.get_scheduler_hint('slo'))
+
+    return not slo_running_request >= slo_new_request
```
Comment (fabiomorais) [inline nova/scheduler/preemption_policy/fcfs_preemption_policy.py]: Change to "return a < b"

File: nova/scheduler/preemption_policy/fcfs_preemption_policy.py [hunk 6]
```diff
@@ -1,21 +1,21 @@
...
-    priority_spec2 = int(spec_obj2.get_scheduler_hint('priority'))
+    slo_spec1 = float(spec_obj1.get_scheduler_hint('slo'))
+    slo_spec2 = float(spec_obj2.get_scheduler_hint('slo'))
 
     submit_time_spec1 = dateutil.parser.parse(spec_obj1.get_scheduler_hint('submit_time'))
     submit_time_spec2 = dateutil.parser.parse(spec_obj2.get_scheduler_hint('submit_time'))
 
-    if priority_spec1 > priority_spec2:
+    if slo_spec1 < slo_spec2:
         return 1
```
Comment (fabiomorais) [inline nova/scheduler/preemption_policy/fcfs_preemption_policy.py]: Change hard coded numbers to be constants and remove unnecessary "else"

File: nova/scheduler/preemption_policy/availability_aware_preemption_policy.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+def _compare_vms(spec_obj1, spec_obj2):
+    """
+    Method that compares two requests considering slo, availability and submit time.
+    """
+
+    slo_spec1 = float(spec_obj1.get_scheduler_hint('slo'))
+    slo_spec2 = float(spec_obj2.get_scheduler_hint('slo'))
+
+    if slo_spec1 < slo_spec2:
+        return 1
```
Comment (fabiomorais) [inline nova/scheduler/preemption_policy/availability_aware_preemption_policy.py]: Change hard coded numbers to be constants and remove unnecessary "else"

File: nova/scheduler/preemption_policy/__init__.py [hunk 8]
```diff
@@ -176,22 +217,28 @@ def get_host_capacity(self, host_state):
...
-    #
-    #     with open(self._status_file, 'wb') as file:
-    #         pickle.dump(status_dict, file, pickle.HIGHEST_PROTOCOL)
+    def _get_policy(self, conf):
+        path = '/opt/stack/nova/nova/scheduler/preemption_policy/'
+        policy_conf = conf._namespace._normalized[0]['DEFAULT']['preemption_policy'][0]
+        policy = imp.load_source(policy_conf, path + policy_conf + '.py')
+        return policy
+
+        # def _read_scheduler_state(self):
```
Comment (fabiomorais) [inline nova/scheduler/preemption_policy/__init__.py]: Remove commented code.

File: nova/scheduler/preemption_policy/__init__.py [hunk 9]
```diff
@@ -1,8 +1,14 @@
-import pickle
+from datetime import datetime
+import imp
+
+import dateutil
 
 import nova.conf
 
-from nova.scheduler.preemption_policy import fcfs_preemption_policy
+import fcfs_preemption_policy
```
Comment (fabiomorais) [inline nova/scheduler/preemption_policy/__init__.py]: Are these imports needed?

File: nova/scheduler/weights/worst_fit.py (hunk 1 above)
Comment (jvmafra) [inline nova/scheduler/weights/worst_fit.py]: Ok. Empty lines between the imports and class definition is a suggestion of PEP8

File: nova/scheduler/weights/best_fit.py (hunk 2 above)
Comment (jvmafra) [inline nova/scheduler/weights/best_fit.py]: Ok. Empty lines between the imports and class definition is a suggestion of PEP8

File: nova/scheduler/weights/best_fit.py (hunk 3 above)
Comment (jvmafra) [inline nova/scheduler/weights/best_fit.py]: Preemption policy calculates free capacity on hosts based on free space and preemptable vms (Following FCFS or availability-aware) on host. It's independent of allocation policy (Best-fit and Worst-fit)

File: nova/scheduler/preemptive_filter_scheduler.py (hunk 4 above)
Comment (jvmafra) [inline nova/scheduler/preemptive_filter_scheduler.py]: Ok

File: nova/scheduler/preemption_policy/fcfs_preemption_policy.py (hunk 5 above)
Comment (jvmafra) [inline nova/scheduler/preemption_policy/fcfs_preemption_policy.py]: Ok

File: nova/scheduler/preemption_policy/__init__.py (hunk 9 above)
Comment (jvmafra) [inline nova/scheduler/preemption_policy/__init__.py]: Yes, they are.

File: nova/scheduler/preemption_policy/__init__.py (hunk 8 above)
Comment (jvmafra) [inline nova/scheduler/preemption_policy/__init__.py]: This commented code will be used in the future

File: nova/scheduler/preemptive_filter_scheduler.py (hunk 4 above)
Comment (fabiomorais) [inline nova/scheduler/preemptive_filter_scheduler.py]: The master code do not use list comprehensions. Example to filter the hosts with max weight: my_list = filter(lambda x: x..weight == max_weight, my_list)

File: nova/scheduler/preemptive_filter_scheduler.py [hunk 10]
```diff
@@ -165,7 +165,7 @@ def _schedule(self, context, spec_obj):
             print "Weighed hosts:"
             print weighed_hosts
             
-            if len(weighed_hosts) > 1:
+            if weighed_hosts:
```
Comment (fabiomorais) [inline nova/scheduler/preemptive_filter_scheduler.py]: @alessandroliafook @jvmafra Please, use list comprehension or lambda filter

https://stackoverflow.com/questions/3013449/list-filtering-list-comprehension-vs-lambda-filter

**Pull request merged**